### ISSUE TITLE
build_proton.sh failed on Intel edison

### ISSUE THREAD

**Issue by krammer0** [Azure/azure-iot-sdks#241]
Hi,

I'm trying azure iot suite on Intel edison.
I followed the instructions on https://github.com/Azure/azure-iot-sdks/blob/master/c/doc/run_sample_on_intel_edison.md, and everything was fine until build_proton.sh.
I tried to change CMake from version 2.8 to 3.4, but the result was the same.

Following are some logs:

root@krammer:~/azure-iot-sdks/c/build_all/linux# ./build_proton.sh 
Source: /home/root/qpid-proton
Install: /home/root
Azure IoT SDK has a dependency on apache qpid-proton-c
https://github.com/apache/qpid-proton/blob/master/LICENSE
Do you want to install the component (y/n)?y
preparing qpid proton-c
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0100   138    0   138    0     0     64      0 --:--:--  0:00:02 --:--:--    71
  0     0    0     0    0     0      0      0 --:--:--  0:00:03 --:--:--     0  0     0    0     0    0     0      0      0 --:--:--  0:00:03 --:--:--     0  5 1663k    5 95219    0     0  20209      0  0:01:24  0:00:04  0:01:20 64120 54 1663k   54  913k    0     0   162k      0  0:00:10  0:00:05  0:00:05  379k100 1663k  100 1663k    0     0   267k      0  0:00:06  0:00:06 --:--:--  557k
Archive:  proton.zip

...skip the inflating...

In build
-- Build type is "RelWithDebInfo" (has debug symbols)
-- The C compiler identification is GNU 4.9.1
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- PN_VERSION: 0.9
-- Could NOT find Java (missing:  Java_JAR_EXECUTABLE Java_JAVAC_EXECUTABLE Java_JAVAH_EXECUTABLE Java_JAVADOC_EXECUTABL
E) 
-- Found OpenSSL: /usr/lib/libssl.so;/lib/libcrypto.so (found version "1.0.1m") 
-- Found PythonInterp: /usr/bin/python (found version "2.7.3") 
-- Looking for clock_gettime
-- Looking for clock_gettime - found
-- Looking for uuid_generate
-- Looking for uuid_generate - not found
-- Looking for uuid_generate in uuid
-- Looking for uuid_generate in uuid - found
-- Looking for strerror_r
-- Looking for strerror_r - found
-- Looking for atoll
-- Looking for atoll - found
-- Found SWIG: /usr/bin/swig (found version "3.0.5") 
-- Found PythonLibs: /usr/lib/libpython2.7.so (found version "2.7.3") 
-- Could NOT find Ruby (missing:  RUBY_EXECUTABLE RUBY_INCLUDE_DIR RUBY_LIBRARY) 
-- Found Perl: /usr/bin/perl (found version "5.20.0") 
Warning: failed to load Config_git.pl, something strange about this perl...
Warning: failed to load Config_git.pl, something strange about this perl...
Warning: failed to load Config_git.pl, something strange about this perl...
Warning: failed to load Config_git.pl, something strange about this perl...
Warning: failed to load Config_git.pl, something strange about this perl...
Warning: failed to load Config_git.pl, something strange about this perl...
Warning: failed to load Config_git.pl, something strange about this perl...
Warning: failed to load Config_git.pl, something strange about this perl...
Warning: failed to load Config_git.pl, something strange about this perl...
Warning: failed to load Config_git.pl, something strange about this perl...
Warning: failed to load Config_git.pl, something strange about this perl...
-- Could NOT find PerlLibs (missing:  PERL_LIBRARY PERL_INCLUDE_PATH) (found version "5.20.0")
-- Trying alternative search for Perl
Warning: failed to load Config_git.pl, something strange about this perl...
-- Could not find perl.h
Warning: failed to load Config_git.pl, something strange about this perl...
-- Looking for libperl.so
Warning: failed to load Config_git.pl, something strange about this perl...
-- No Perl devel environment found - skipping Perl bindings
-- Emscripten (https://github.com/kripken/emscripten) is not installed: can't build JavaScript binding
-- Could NOT find Doxygen (missing:  DOXYGEN_EXECUTABLE) 
-- Looking for include file inttypes.h
-- Looking for include file inttypes.h - found
-- Can't locate the valgrind command; no run-time error detection
-- Cannot find ruby, skipping ruby tests
-- Cannot find both Java and Maven: testing disabled for Proton-J
-- Configuring done
-- Generating done
-- Build files have been written to: /home/root/qpid-proton/build
[  1%] Generating src/protocol.h
[  3%] Generating src/encodings.h
Scanning dependencies of target qpid-proton
[  4%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/object/object.c.o
[  6%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/object/list.c.o
[  7%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/object/map.c.o
[  9%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/object/string.c.o
[ 10%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/object/iterator.c.o
[ 12%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/object/record.c.o
[ 13%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/log.c.o
[ 15%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/util.c.o
[ 16%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/url.c.o
[ 18%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/error.c.o
[ 19%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/buffer.c.o
[ 21%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/parser.c.o
[ 22%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/scanner.c.o
[ 24%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/types.c.o
[ 25%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/framing/framing.c.o
[ 27%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/codec/codec.c.o
[ 28%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/codec/decoder.c.o
[ 30%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/codec/encoder.c.o
[ 31%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/dispatcher/dispatcher.c.o
[ 33%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/engine/engine.c.o
[ 34%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/events/event.c.o
[ 36%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/transport/autodetect.c.o
[ 37%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/transport/transport.c.o
[ 39%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/message/message.c.o
[ 40%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/sasl/sasl.c.o
[ 42%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/reactor/reactor.c.o
[ 43%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/reactor/handler.c.o
[ 45%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/reactor/connection.c.o
[ 46%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/reactor/acceptor.c.o
[ 48%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/reactor/timer.c.o
[ 50%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/handlers/handshaker.c.o
[ 51%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/handlers/iohandler.c.o
[ 53%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/handlers/flowcontroller.c.o
[ 54%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/messenger/messenger.c.o
[ 56%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/messenger/subscription.c.o
[ 57%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/messenger/store.c.o
[ 59%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/messenger/transform.c.o
[ 60%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/selectable.c.o
[ 62%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/posix/io.c.o
[ 63%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/posix/selector.c.o
[ 65%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/platform.c.o
[ 66%] Building C object proton-c/CMakeFiles/qpid-proton.dir/src/ssl/openssl.c.o
Linking C shared library libqpid-proton.so
[ 66%] Built target qpid-proton
Scanning dependencies of target proton-dump
[ 68%] Building C object proton-c/CMakeFiles/proton-dump.dir/src/proton-dump.c.o
Linking C executable proton-dump
[ 68%] Built target proton-dump
[ 69%] Swig source
Scanning dependencies of target _cproton
[ 71%] Building C object proton-c/bindings/python/CMakeFiles/_cproton.dir/cprotonPYTHON_wrap.c.o
Linking C shared module _cproton.so
CMakeFiles/_cproton.dir/cprotonPYTHON_wrap.c.o: In function `_wrap_pn_messenger_set_client_sasl_mechanism':
/home/root/qpid-proton/build/proton-c/bindings/python/cprotonPYTHON_wrap.c:19630: undefined reference to `pn_messenger_s
et_client_sasl_mechanism'
collect2: error: ld returned 1 exit status
proton-c/bindings/python/CMakeFiles/_cproton.dir/build.make:129: recipe for target 'proton-c/bindings/python/_cproton.so
' failed
make[2]: *** [proton-c/bindings/python/_cproton.so] Error 1
CMakeFiles/Makefile2:1094: recipe for target 'proton-c/bindings/python/CMakeFiles/_cproton.dir/all' failed
make[1]: *** [proton-c/bindings/python/CMakeFiles/_cproton.dir/all] Error 2
Makefile:123: recipe for target 'all' failed
make: *** [all] Error 2

Did I miss something? Thanks in advance.

Thanks,
Krammer

**Comment by tameraw** [comment]
@krammer0 - Thanks for reporting this issue. We're looking at it and will update you.
Regards.

**Comment by ashokkhurana** [comment]
Hi Krammer,

Could you please try the instructions available in our develop branch at https://github.com/Azure/azure-iot-sdks/blob/develop/doc/get_started/yocto-intel-edison-c.md . These are the updated instructions for Intel Edison and will take care of the issue you are getting. The master branch document is old and will be updated in our next release.

Thanks,
Ashok

**Comment by krammer0** [comment]
Hi Ashok,

After step 5 and 6, and I download azure-c-shared-utility manually, it works!

But when it went to test, I got so many error.

Constructing a list of tests
Done constructing a list of tests
Checking test dependency graph...
Checking test dependency graph end
test 1
      Start  1: ctest_unittests

1: Test command: /home/root/cmake/azure-c-shared-utility/c/testtools/ctest/unittests/ctestunittests/ctest_unittests
1: Test timeout computed to be: 9.99988e+06
1:  === Executing test suite SimpleTestSuiteOneTest ===
1: Executing test Test1 ...
1: Suceeded.
1: 1 tests ran, 0 failed, 1 succeeded.
1:  === Executing test suite SimpleTestSuiteOneTest ===
1: Executing test Test1 ...
1: Suceeded.
1: 1 tests ran, 0 failed, 1 succeeded.
1:  === Executing test suite SimpleTestSuiteTwoTests ===
1: Executing test Test2 ...
1: Suceeded.
1: Executing test Test1 ...
1: Suceeded.
1: 2 tests ran, 0 failed, 2 succeeded.
1:  === Executing test suite AssertFailureTests ===
1: Executing test Assert_Is_False_With_Msg_Fails ...
1:   Assert failed: Expression should be false: 1. Test message
1: !!! FAILED !!!
1: Executing test Assert_Is_False_Fails ...
1:   Assert failed: Expression should be false: 1. 
1: !!! FAILED !!!

...skip some error...

52/52 Test #52: serializer_e2etests ...................***Failed    0.04 sec

94% tests passed, 3 tests failed out of 52

Total Test time (real) =  85.15 sec

The following tests FAILED:
	 31 - iothubclient_amqp_e2etests (Failed)
	 33 - iothubclient_http_e2etests (Failed)
	 52 - serializer_e2etests (Failed)
Errors while running CTest

And then when I run iothub_client_sample_amqp I got

root@krammer:~/cmake/iothub_client/samples/iothub_client_sample_amqp# ./iothub_cllient_sample_amqp 
Starting the IoTHub client sample AMQP...
Info: IoT Hub SDK for C, version 1.0.0-preview.9
Error: Time:Tue Jan 26 08:34:29 2016 File:/home/root/azure-iot-sdks/c/iothub_client/src/iothub_client_ll.c Func:IoTHubCl
ient_LL_CreateFromConnectionString Line:119 Tokenizer error
Error: Time:Tue Jan 26 08:34:29 2016 File:/home/root/azure-iot-sdks/c/iothub_client/src/iothub_client_ll.c Func:IoTHubCl
ient_LL_CreateFromConnectionString Line:194 iotHubName is not found
Error: Time:Tue Jan 26 08:34:29 2016 File:/home/root/azure-iot-sdks/c/iothub_client/src/iothub_client.c Func:IoTHubClien
t_CreateFromConnectionString Line:106 IoTHubClient_LL_CreateFromConnectionString failed
ERROR: iotHubClientHandle is NULL!

Should I open another issue?

Thanks,
Krammer

**Comment by ashokkhurana** [comment]
Hi Krammer,

E2E test cases are failing as they need few environment variables been set on the board. Please refer https://github.com/Azure/azure-iot-sdks/blob/master/c/doc/run_end_to_end_tests.md to see what necessary environment variable need to be in place for E2E.

Issue you are getting while running sample can be due to the connection string either not provided or a wrong value is given in file "/c/iothub_client/samples/iothub_client_sample_amqp/iothub_client_sample_amqp.c". Could you please check that device connection string is properly set in this file? You can get this connection string from device explorer by right clicking your device and using option "Copy Connection String for Device".

Please let us know how it goes.

Thanks,
Ashok

**Comment by krammer0** [comment]
Hi Ashok,

After applying the connection string from device explorer, I can run the iothub_client_sample_amqp successfully.
Thanks for your big help!

Regards,
Krammer

**Comment by tameraw** [comment]
@krammer0 - Thanks for the update.
